### PR TITLE
internal/repos: Implement connectionCheck

### PR DESCRIPTION
**Pull request by indradhanush**
Part of #44683. 

This commit adds a connection check for the following code hosts:

1. Bitbucket cloud
2. Bitbucket server
3. GitHub
4. Gitlab

## Test plan

1. Added tests
2. Tested locally
3. Build should pass
4. [main-dry-run](https://github.com/sourcegraph/sourcegraph/tree/main-dry-run/ig/connection-check) should pass

**Connection check fails**
image

**Connection check passes**
image

### CODE REVIEW

File: internal/repos/check_connection.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
+package repos
+
+import (
+	"net"
+	"net/url"
+
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+)
+
+// checkConnection will parse the rawURL and make a best effort attempt to obtain a hostname. It
```
Comment (sashaostrikov) [inline internal/repos/check_connection.go:10]: nit: usually docs are written in present simple
```suggestion
// checkConnection parses the rawURL and makes a best effort attempt to obtain a hostname. It
```

File: internal/repos/check_connection.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+// then performs an IP lookup on that hostname and will return a non-nil error on failure.
+//
+// At the moment this function is only limited to doing IP lookups. We may want / have to expand
```
Comment (sashaostrikov) [inline internal/repos/check_connection.go:13]:
```suggestion
// At the moment this function is only limited to doing IP lookups. We may want/have to expand
```

File: internal/repos/check_connection.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+// this to support other code hosts or to add more checks (example making a test API call to verify
```
Comment (sashaostrikov) [inline internal/repos/check_connection.go:14]:
```suggestion
// this to support other code hosts or to add more checks (e.g. making a test API call to verify
```

File: internal/repos/check_connection.go [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+// the authorization etc)
```
Comment (sashaostrikov) [inline internal/repos/check_connection.go:15]:
```suggestion
// the authorization, etc.).
```

File: internal/repos/check_connection.go [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+func checkConnection(rawURL string) error {
+	u, err := url.Parse(rawURL)
+	if err != nil {
+		return errors.Wrap(err, "connection check failed")
+	}
+
+	// Best effort at finding a hostname. For example if rawURL is sourcegraph.com, then u.Host is
+	// empty but Path is sourcegraph.com. Use that as a result.
+	//
+	// 👉 Also, we need to use u.Hostname() here because we want to strip any port numbers if they
```
Comment (sashaostrikov) [inline internal/repos/check_connection.go:25]: why the emoji? 😀

File: internal/repos/check_connection.go [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+	// are present in u.Host.
+	hostname := u.Hostname()
+	if hostname == "" {
+		hostname = u.Path
+	}
+
+	ips, err := net.LookupIP(hostname)
+	if err != nil {
+		return errors.Wrap(err, "connection check failed")
```
Comment (ryanslade) [inline internal/repos/check_connection.go:34]: both errors are wrapped with the same text which will make it difficult to differentiate them.

I think the first one should mention there was an error parsing the url and the second should mention there was an error looking up the IP.

File: internal/repos/check_connection_test.go [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+		}
+	})
+
+	t.Run("good URL with port", func(t *testing.T) {
+		if err := checkConnection("https://sourcegraph.com:80"); err != nil {
+			t.Errorf("Expected nil but got error: %v", err)
+		}
+	})
+
+	t.Run("good URL without protocol", func(t *testing.T) {
```
Comment (sashaostrikov) [inline internal/repos/check_connection_test.go:24]: should we add a test case "without protocol with port"?

File: internal/repos/check_connection_test.go [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+		if err := checkConnection("sourcegraph.com"); err != nil {
+			t.Errorf("Expected nil but got error: %v", err)
+		}
+	})
+	t.Run("good URL with username:password", func(t *testing.T) {
```
Comment (ryanslade) [inline internal/repos/check_connection_test.go:29]: super nit: No spacing between these tests, but others do have spacing.

Comment (indradhanush) [reply internal/repos/check_connection_test.go:29]: I like this nit. 😁

File: internal/repos/check_connection_test.go [hunk 9]
```diff
@@ -0,0 +1,34 @@
+package repos
+
+import "testing"
+
+func TestCheckConnection(t *testing.T) {
+	t.Run("bad URL", func(t *testing.T) {
```
Comment (sashaostrikov) [inline internal/repos/check_connection_test.go:6]: should we use a testcase slice pattern so we don't repeat the `checkConnection` calling code (it's short, but still)

**Review by ryanslade — APPROVED** [review]

**Review by sashaostrikov — APPROVED** [review]
LGTM overall!
I've added some docs and testing suggestions :)

File: internal/repos/check_connection_test.go (hunk 8 above)
Comment (sashaostrikov) [reply internal/repos/check_connection_test.go:30]: @ryanslade lol, I thought the same, but decided to skip this comment :) teamwork!

File: internal/repos/check_connection.go (hunk 1 above)
Comment (indradhanush) [reply internal/repos/check_connection.go:10]: I prefer it in its current form.

Google's technical writing guide suggests to use active voice over passive voice: https://developers.google.com/tech-writing/one/active-voice.

File: internal/repos/check_connection.go (hunk 5 above)
Comment (indradhanush) [reply internal/repos/check_connection.go:25]: To get the attention of someone reading this code. It worked didn't it? 😝

File: internal/repos/check_connection_test.go (hunk 9 above)
Comment (indradhanush) [reply internal/repos/check_connection_test.go:6]: For simpler test cases, I like the independent tests. The table driven adds more code for adding a struct for expected result and makes it hard to read IMO.

File: internal/repos/check_connection.go (hunk 1 above)
Comment (sashaostrikov) [reply internal/repos/check_connection.go:10]: I mostly mean that docs with present simple are more frequent in our code base and it's better to keep to one or more popular thing for the sake of uniformity :)

Comment (indradhanush) [reply internal/repos/check_connection.go:10]: I disagree. 😁

Active voice should be preferred and we should attempt to move towards that instead of continuing to stick to using passive voice. 

Active voice is simpler and more effective for the reasons mentioned in the guide I linked above.

File: internal/repos/check_connection.go (hunk 2 above)
Comment (sashaostrikov) [reply internal/repos/check_connection.go:13]: I'm not a huge punctuation professor in this case, so I'll leave it up to you :)

File: internal/repos/check_connection.go (hunk 1 above)
Comment (mrnugget) [reply internal/repos/check_connection.go:10]: I think the official Go docs overrule the Google tech writing guideline: https://go.dev/blog/godoc

Also: `checkConnection prases the rawURL` is still an active voice. The difference is the future vs. present tense.

File: internal/repos/check_connection.go (hunk 5 above)
Comment (sashaostrikov) [reply internal/repos/check_connection.go:25]: I usually read all of the doc, if I read it, should be fine by itself.
Wrapping some terms/names with quotes, on the other hand, makes reading a bit easier for me, as in function names or path example (`sourcegraph.com`)

**Review by mrnugget — APPROVED** [review]
This does `net.LookupIP`, which does a DNS resolve, as far as I know, right? It doesn't actually check whether a connection can be established, correct?

Say GitHub is down, for example, but the hostname is resolvable, what happens then?

File: internal/repos/check_connection.go (hunk 1 above)
Comment (mrnugget) [reply internal/repos/check_connection.go:10]: Yeah, I was referring to Indra's comment 

> Google's technical writing guide suggests to use active voice over passive voice:

Comment (indradhanush) [reply internal/repos/check_connection.go]: Makes sense. Fixed.

File: internal/repos/check_connection.go (hunk 2 above)
Comment (indradhanush) [reply internal/repos/check_connection.go]: Fixed. 😀

File: internal/repos/check_connection_test.go (hunk 7 above)
Comment (indradhanush) [reply internal/repos/check_connection_test.go:24]: Great suggestion. Tests were failing. Reworked the code and added the test.

**Comment by indradhanush** [comment]
Requested another review because the code changed a bit in an ugly way and reduces my confidence somewhat because of the new edge case of `sourcegraph.com:80` was making it fail. This is brittler than I thought.

**Comment by indradhanush** [comment]
> This does net.LookupIP, which does a DNS resolve, as far as I know, right? It doesn't actually check whether a connection can be established, correct?

Correct. 

> Say GitHub is down, for example, but the hostname is resolvable, what happens then?

Great question. I think we should attempt at making a test API call that preferably won't consume rate limit quota (like reading rate-limit info from the GitHub API or the version API request).

This will vary more than the current implementation for each code host and maybe can be worked in as an after thought of improving the connection checks once we have the feature in place? What do you think?

File: internal/repos/check_connection.go [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+		if u.Scheme != "" {
+			// rawURL is most likely something like "sourcegraph.com:80", read from u.Scheme.
+			hostname = u.Scheme
+		} else if u.Path != "" {
+			// rawURL is most likely something like "sourcegraph.com:80", read from u.Path.
+			hostname = u.Path
+		} else {
+			return errors.Newf("unsupported url format (%q) for connection check", rawURL)
+		}
+	}
```
Comment (mrnugget) [inline internal/repos/check_connection.go:38]: For Bitbucket Server/Bitbucket Cloud/GitHub we already validate that the URL in the config is a proper URL:

- https://sourcegraph.sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/schema/bitbucket_server.schema.json?L48:5
- https://sourcegraph.sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/schema/bitbucket_cloud.schema.json?L19
- https://sourcegraph.sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/schema/gitlab.schema.json?L11-20
- https://sourcegraph.sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/schema/github.schema.json?L11-20

I think that should help be a bit more confident here.

**Comment by sashaostrikov** [comment]
@indradhanush can we replace dns check with a ping? I think it can solve the problem and be a relatively simple thing to add right away

**Comment by mrnugget** [comment]
> This will vary more than the current implementation for each code host and maybe can be worked in as an after thought of improving the connection checks once we have the feature in place? What do you think?

Yeah, fine with me to do this in a follow-up PR, but I think this is not an "after thought" but was part of the original idea :) See also: https://github.com/sourcegraph/sourcegraph/pull/44922#discussion_r1038317601

**Comment by mrnugget** [comment]
> @indradhanush can we replace dns check with a ping? I think it can solve the problem and be a relatively simple thing to add right away

DNS check is still useful, though. It's just one part of it.

**Review by sashaostrikov — APPROVED** [review]

**Comment by indradhanush** [comment]
> Yeah, fine with me to do this in a follow-up PR, but I think this is not an "after thought" but was part of the original idea :) See also: https://github.com/sourcegraph/sourcegraph/pull/44922#discussion_r1038317601

Aha! That is useful context. I will follow up in another PR then.

File: internal/repos/check_connection.go (hunk 10 above)
Comment (indradhanush) [reply internal/repos/check_connection.go:38]: Nice! 😍

Comment (mrnugget) [reply internal/repos/check_connection.go:38]: To follow-up: did you change the code to take that into account?

Comment (indradhanush) [reply internal/repos/check_connection.go:38]: We chatted on Slack, but leaving this here for posterity's sake:

With the checks in the schema it did add confidence when I was implementing it. But I felt it prudent to err on the side of caution and build in the checks for empty scheme / path anyway. It somehow felt wrong to take `rawURL` as some sort of a guaranteed format, especially if someone else uses it for another purpose down the road.
